xls/contrib/integrator/ir_integrator.cc

```diff
@@ -194,6 +244,168 @@ absl::StatusOr<Node*> IntegrationFunction::InsertNode(const Node* to_insert) {
   return inserted;
 }
 
+absl::StatusOr<IntegrationFunction::MergeNodesBackendResult>
+IntegrationFunction::MergeNodesBackend(const Node* node_a, const Node* node_b) {
+  auto validate_node = [this](const Node* node) {
+    if (IntegrationFunctionOwnsNode(node)) {
+      if (!IsMappingTarget(node)) {
+        return absl::InternalError(
+            "Trying to merge non-mapping-target integration node.");
+      }
+    } else {
+      if (HasMapping(node)) {
+        return absl::InternalError(
+            "Trying to merge non-integration node that already has mapping");
+      }
+    }
+    return absl::OkStatus();
+  };
+  XLS_RETURN_IF_ERROR(validate_node(node_a));
+  XLS_RETURN_IF_ERROR(validate_node(node_b));
+  // Simple way to avoid situation where node_a depends on node_b or
+  // vice versa. If we later want to merge nodes from the same function,
+  // we can implement reaching analysis to check for dependencies.
+  if (node_a != node_b) {
+    XLS_RET_CHECK(node_a->function_base() != node_b->function_base());
+  }
+
+  // Separate targets for node_a and node_b because we may derive
+  // map target nodes from the merged node;
+
+  // Identical nodes can always be merged.
+  if (node_a->IsDefinitelyEqualTo(node_b)) {
+    XLS_ASSIGN_OR_RETURN(UnifiedOperands unified_operands,
+                         UnifyNodeOperands(node_a, node_b));
+    XLS_ASSIGN_OR_RETURN(
+        Node * merged,
+        node_a->CloneInNewFunction(unified_operands.operands, function()));
+    MergeNodesBackendResult merge_result;
+    merge_result.can_merge = true;
+    merge_result.target_a = merged;
+    merge_result.target_b = merged;
+    merge_result.added_muxes = unified_operands.added_muxes;
+    merge_result.other_added_nodes.insert(merged);
+    return merge_result;
+  } else {
+    // TODO(jbaileyhandle): Add logic for nodes that are not identical but
+    // may still be merged e.g. different bitwidths for bitwise ops.
+    switch (node_a->op()) {
+      default:
+        break;
+    }
+  }
+
+  MergeNodesBackendResult merge_result;
+  merge_result.can_merge = false;
+  return merge_result;
+}
+
+absl::StatusOr<std::optional<float>> IntegrationFunction::GetMergeNodesCost(
+    const Node* node_a, const Node* node_b) {
+  XLS_ASSIGN_OR_RETURN(MergeNodesBackendResult merge_result,
+                       MergeNodesBackend(node_a, node_b));
+  // Can't merge nodes.
+  if (!merge_result.can_merge) {
+    return absl::nullopt;
+  }
+
+  // Score.
+  float cost = 0.0;
+  auto tabulate_node_elimination_cost = [this, &cost](const Node* node) {
+    if (IntegrationFunctionOwnsNode(node)) {
+      cost -= GetNodeCost(node);
+    }
+  };
+  tabulate_node_elimination_cost(node_a);
+  tabulate_node_elimination_cost(node_b);
+  for (Node* new_node : merge_result.added_muxes) {
+    cost += GetNodeCost(new_node);
+  }
+  for (Node* new_node : merge_result.other_added_nodes) {
+    cost += GetNodeCost(new_node);
+  }
+
+  // Cleanup.
```

Comment:
Is there a way to compute the cost without mutating the graph? It is a bit surprising that a method called GetMergeNodeCost adds nodes to the graph then removes them.

Reply:
I think it's the lesser of two evils. I had a couple of things in mind that motivated this decision:

-We'll want to add different ways to merge non-identical nodes (simple example: merge adds of different bitwidths, then use a bitslice to create a node for the narrower of the two merged nodes, implement.  Other non-identical merges may involve adding other nodes that are more costly than bitslice / can't be assumed to be free).  If we have a non-mutating GetMergeNodesCost, we'll have to duplicate most of the logic for merging and for getting the cost of a merge (e.g. we'll have "do this" and "what would I do" version of all merge code).  This is both a waste of effort and is likely to introduce errors, for instance if we implement the merge + cost versions of a techinque but then later update only the merge version.

-Similar to the above, the do vs. estimate split would likely bleed out to code beyond just that directly implementing merging.  For instance, we have functions to mux together nodes that are used on operands when merging nodes.  This, too, would need to be split into two versions.  Further, it could be difficult to do this correctly.  For instance, if we need to mux together the same two nodes for multiple purposes, we reuse the same mux.  If we are trying to estimate the cost of merging two nodes where we have two operands that will mux together the same nodes (e.g merge (Node(a,a) and Node(b,b) -> Node(mux(a,b, mux(a,b)), we'd need to only count the cost of adding one mux.  This gets tricky - currently, GetNodesMergeCost outsources the operand muxing to UnifyNodeOperands, which outsources per operand to UnifyIntegrationNodes. UnifyIntegrationNodes can checks some bookkeping to see if we've made this mux before.  For cost-only functions, we haven't modified this global book-keeping, so we'd have to handle duplicate checking higher up in the call stack.  Said another way, helper functions might have to be collapsed into cost estimation functions.

-Modularity would likely be hurt.  There's the helper function collapsing issue mention above.  Beyond that, we'd have to keep the the merge vs cost split in mind when designing other code.  There may be solutions to the merge vs cost issues we see in the code now, but I'm worried those solutions would break if assumptions or other design components change e.g. merge techniques only add 1 costly node and the others are trivial things like bit slices or oncats at some point in developement, so we decide we only have to id the coslty node, but this breaks if we add more complicated merge techniques. Or similarly, perhaps we make assumptions about how we add muxes or inline functions or track node mappings or any other number of things at some point, but we may change these later.  The unkown uknowns seem to present a lot of risk here.

In summary, while I agree that having a non-mutating cost estimate is conceptually cleaner, I think this would likely lead to a lot of pain and suffering on the implementation end and open up possibilies for bugs.  Another approach to estimating cost without mutating the graph is to clone the function, estimate the cost of the whole graph, apply the merge on the cloned function, and then see how much the cost changed.  However, this seems very inefficient.

Reply:
sounds reasonable. Thanks for the explanation. My only suggestion then is to make it as simple as possible to rollback the nodes that were added to minimize the possibility of leaving stuff in the graph.